### PR TITLE
Add interpreter for MapOp

### LINKED ISSUE #1106: Add  interpreter for MapOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) inputs: variadic number of tensors.
(I2) dimensions: 1-dimensional tensor constant of type `si64`.
(I3) computation: function.
(C1) All `inputs` and `result` have the same shape.
(C2) size(`inputs`) = N >= 1.
(C3) `dimensions = [0, ..., R-1]`, where `R` = rank(`inputs[0]`).
(C4) `computation` has type `(tensor<E0>, ..., tensor<EN-1>) -> tensor<E'>`
where `Ek` = element_type(`inputs[k]`) and `E'` = element_type(`result`).
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) inputs is not a variadic tensor. (Covered by ODS).
I2: a) dimensions is not a 1-dimensional tensor.
    b) element_type(dimensions) != si64. (Covered by ODS).
I3: a) computation is not a function. (Covered by ODS).
C1: a) Not all `inputs` have the same shape. (Covered by ODS).
    b) `inputs` and `result` do not have the same shape. (Covered by ODS).
C2: size(`inputs`) < 1. (Covered by ODS).
C3: a) `dimensions != [0, ..., R-1]`, where `R` = rank(`inputs[0]`).
C4: a) `computation` does not have type `(tensor<E0>, ..., tensor<EN-1>) -> tensor<E'>`
where `Ek` = element_type(`inputs[k]`) and `E'` = element_type(`result`).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I2a: dimensions is not a 1-dimensional tensor.
C3a: `dimensions != [0, ..., R-1]`, where `R` = rank(`inputs[0]`).
C4a: `computation` does not have type `(tensor<E0>, ..., tensor<EN-1>) -> tensor<E'>`
where `Ek` = element_type(`inputs[k]`) and `E'` = element_type(`result`).
```

closes #1106

### CODE REVIEW

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -2354,6 +2354,12 @@ LogicalResult inferMapOp(
     std::optional<Location> location, ValueRange inputs,
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:2354]: Missing labels for constraints like C3a and c4a.

File: stablehlo/dialect/StablehloOps.td [hunk 2]
```diff
@@ -2371,7 +2371,8 @@ def StableHLO_GetDimensionSizeOp: StableHLO_Op<"get_dimension_size",
 }
 
 def StableHLO_MapOp: StableHLO_ShapedInterfaceOp<"map",
-      [RecursiveMemoryEffects, SameOperandsAndResultShape,
+      [RecursiveMemoryEffects, SameVariadicOperandSize /*map_c2*/,
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:2374]: No need to add `SameVariadicOperandSize` if we do not have multiple variadic operands?

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -2354,6 +2354,12 @@ LogicalResult inferMapOp(
...
     DenseIntElementsAttr dimensions, Region& computation,
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  // map_i2
+  if (dimensions.getType().getRank() != 1)
+    return emitOptionalError(location,
+                             "dimensions should be rank 1 but got rank ",
+                             dimensions.getType().getRank());
+
   if (failed(verifyRegionNotEmpty(location, computation))) return failure();
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:2363]: Also add a check for c2 similar to what HLO has
https://github.com/tensorflow/tensorflow/blob/4f1dd6d5123f4eb6afc85fac36df09b4a8b49c83/tensorflow/compiler/xla/service/shape_inference.cc#L1161

File: stablehlo/tests/ops_stablehlo.mlir [hunk 4]
```diff
@@ -1610,54 +1610,43 @@ func.func @map(%arg0: tensor<4x5xf32>, %arg1: tensor<4x5xf32>) -> tensor<4x5xf32
...
-    "stablehlo.return"(%arg2) : (tensor<f32>) -> ()
-  }) {dimensions = dense<0> : tensor<1xi64>} : (tensor<2xf32>, tensor<2xi32>) -> tensor<2xf32>
-  func.return %0 : tensor<2xf32>
-}
-
-// -----
-
-// CHECK-LABEL: func @map_scalar_operands
-func.func @map_scalar_operands(%arg0: tensor<f32>, %arg1: tensor<f32>) -> tensor<f32> {
+func.func @map_c3(%arg0: tensor<4x5xf32>, %arg1: tensor<4x5xf32>) -> tensor<4x5xf32> {
```
Comment (sdasgup3) [inline stablehlo/tests/ops_stablehlo.mlir:1613]: add a test for c2

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -541,6 +546,23 @@ Tensor evalLogisticOp(const Tensor &operand, TensorType resultType) {
   return result;
 }
 
+Tensor evalMapOp(ArrayRef<Tensor> inputs, Region &computation, Scope &scope,
+                 TensorType resultType) {
+  Tensor result(resultType);
+  for (auto inputIt = inputs[0].index_begin(); inputIt != inputs[0].index_end();
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:552]: IMO this top level iterator is used only to get the result index. In that case, let it be a iterator on result.

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -541,6 +546,23 @@ Tensor evalLogisticOp(const Tensor &operand, TensorType resultType) {
...
+       ++inputIt) {
+    SmallVector<Tensor> args;
+    for (size_t i = 0; i < inputs.size(); ++i) {
+      auto tensor =
```
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:556]: No need to extract an tensor element to get the element type of the tensor.
```
Tensor arg(RankedTensorType::get({}, inputs[i].getElementType());
```

Even better: IIRC there is a PR where we talked about introducing a constructor like 
```
 explicit Tensor(TensorType, const Element &);
```

We should use it here as well.

Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:556]: Thanks for the suggestion. I also recall discussing this somewhere but haven't find which PR it was (unfortunately GH doesn't have a great search engine).

File: stablehlo/dialect/StablehloOps.td (hunk 2 above)
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.td:2374]: `SameVariadicOperandSize` should be renamed to `VariadicOperandsHaveSameSize`. The singular `Operand` is misleading :)

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:2363]: The check is not needed as this is still caught by the ODS with the `verifyAtLeastNOperands(op, 1)` check.
Reference: https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/Operation.cpp#L844-L845

File: stablehlo/tests/ops_stablehlo.mlir (hunk 4 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:1613]: Since the ODS covers this, a test won't be needed.

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp [hunk 7]
```diff
@@ -2387,24 +2391,22 @@ LogicalResult inferMapOp(
...
+  // map_c4
   auto computationOutputs = computationBlock.getTerminator()->getOperands();
   if (computationOutputs.size() != 1)
     return emitOptionalError(location,
                              "computation must return single output, but got: ",
                              computationOutputs.size());
 
-  // The output of computation must be scalar and have the same element type
-  // as op result.
+  // map_c4
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:2401]: Looks like lines 2438-2442 are also `map_c4`?

File: stablehlo/tests/infer_stablehlo.mlir [hunk 8]
```diff
@@ -366,6 +366,18 @@ func.func @map(%arg0: tensor<4x5xf32>, %arg1: tensor<4x5xf32>) -> tensor<4x5xind
...
+  // expected-error@+1 {{inferred type(s) 'tensor<4x5xi32>' are incompatible with return type(s) of operation 'tensor<4x5xf32>'}}
+  %0 = "stablehlo.map"(%arg0, %arg1) ({
+    ^bb0(%arg2: tensor<f32>, %arg3: tensor<f32>):
+    %1 = stablehlo.constant dense<2> : tensor<i32>
+    "stablehlo.return"(%1) : (tensor<i32>) -> ()
+  }) {dimensions = dense<[0, 1]> : tensor<2xi64>} : (tensor<4x5xf32>, tensor<4x5xf32>) -> tensor<4x5xf32>
+  func.return %0 : tensor<4x5xf32>
+}
+
+// -----
```
Comment (burmako) [inline stablehlo/tests/infer_stablehlo.mlir:379]: Let's drop this test, similarly to other PRs that we discussed recently?

File: stablehlo/reference/Ops.cpp [hunk 9]
```diff
@@ -541,6 +546,21 @@ Tensor evalLogisticOp(const Tensor &operand, TensorType resultType) {
...
 
+Tensor evalMapOp(ArrayRef<Tensor> inputs, Region &computation, Scope &scope,
+                 TensorType resultType) {
+  Tensor result(resultType);
+  for (auto resultIt = inputs[0].index_begin();
+       resultIt != inputs[0].index_end(); ++resultIt) {
+    SmallVector<Tensor> args;
+    for (size_t i = 0; i < inputs.size(); ++i)
+      args.push_back(
+          Tensor(RankedTensorType::get({}, inputs[i].get(*resultIt).getType()),
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:557]: Can we use `computation.getArgument(i).getType()` instead? Shorter to type and doesn't require creating types, i.e. avoids potential logic errors.

File: stablehlo/reference/Tensor.cpp [hunk 10]
```diff
@@ -93,6 +93,11 @@ Tensor::Tensor(TensorType type)
 Tensor::Tensor(TensorType type, AsmResourceBlob blob)
     : impl_(llvm::makeIntrusiveRefCnt<detail::Buffer>(type, std::move(blob))) {}
 
+Tensor::Tensor(TensorType type, const Element &element)
+    : impl_(llvm::makeIntrusiveRefCnt<detail::Buffer>(type)) {
+  this->set({}, element);
+}
```
Comment (burmako) [inline stablehlo/reference/Tensor.cpp:99]: I see how this constructor is more convenient that first creating a tensor, and then doing a set. 

However, I think that it's currently not pulling its weight. On the one hand, it's used only once, and its convenience is marginal (compresses a 2-liner into a 1-liner). On the other hand, it does introduce a new function and complicate the API of `Tensor`, so the cost is non-trivial.

Let's inline the constructor into the callsite?

**Review by burmako — CHANGES_REQUESTED** [review]
LGTM modulo a few minor comments.

**Review by burmako — APPROVED** [review]

**Pull request merged**